htlcswitch/iterator.go

```diff
@@ -85,21 +89,38 @@ type HopIterator interface {
 	// EncodeNextHop encodes the onion packet destined for the next hop
 	// into the passed io.Writer.
 	EncodeNextHop(w io.Writer) error
+
+	// ExtractErrorEncrypter returns the ErrorEncrypter needed for this hop,
+	// along with a failure code to signal if the decoding was successful.
+	ExtractErrorEncrypter(ErrorEncrypterExtracter) (ErrorEncrypter,
```

Comment:
This still feels very complicated and has a lot of indirection. Any thoughts on just exposing a `ErrorEncrypter() ErrorEncrypter` method which is a getter as I mentioned [here](https://github.com/lightningnetwork/lnd/pull/775#discussion_r171874271)?

Reply:
@jimpo I agree about the roundaboutness of this approach. I started trying to make it more of a getter, though it breaks some of the interfaces we have set up. Looks like it might be more work than necessary at this time. I would like to circle back and improve this though